File: web/src/domain/ufReajuste.ts
import { HUNDRED, ONE, ZERO, divScale, money, scale, type Money } from "./money.ts";
import { daysBetween } from "./dates.ts";
import type { ReajusteResult, UfValue } from "./models.ts";

/**
 * Restates an amount between two dates through the UF.
 *
 * This is the mechanism Chilean contracts, rents and debts are actually
 * re-adjusted with, and because the UF is published every calendar day it
 * answers at day precision. The UF tracks the CPI with the lag it is built
 * with, so the variation it reports is inflation as the UF carries it, not the
 * INE's month-on-month figure.
 */
const DAYS_IN_YEAR = 365.25;

export const convert = (amount: Money, from: UfValue, to: UfValue): ReajusteResult => {
  if (from.value.cmp(0) <= 0 || to.value.cmp(0) <= 0) {
    throw new Error("El valor de la UF debe ser positivo");
  }

  const units = divScale(amount, from.value, 4);
  const factor = to.value.div(from.value);
  const days = daysBetween(from.date, to.date);

  let annualised = ZERO;
  if (days !== 0) {
    const years = days / DAYS_IN_YEAR;
    const a = Math.pow(Number(factor.toString()), 1 / years) - 1;
    if (Number.isFinite(a)) annualised = scale(money((a * 100).toString()), 2);
  }

  return {
    amount,
    from: from.date,
    to: to.date,
    ufAtFrom: from.value,
    ufAtTo: to.value,
    ufUnits: units,
    // Rounded once, at the end: converting through UF units and rounding twice
    // would drift on large amounts.
    adjustedAmount: scale(amount.times(factor), 0),
    factor: scale(factor, 6),
    variationPct: scale(factor.minus(ONE).times(HUNDRED), 2),
    annualisedPct: annualised,
    days,
  };
};

File: app/src/main/java/cl/tickers/app/domain/engine/UfReajuste.kt
package cl.tickers.app.domain.engine

import cl.tickers.app.domain.model.ReajusteResult
import cl.tickers.app.domain.model.DatedValue
import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode
import java.time.temporal.ChronoUnit

/**
 * Restates an amount between two dates through the UF.
 *
 * This is the mechanism Chilean contracts, rents and debts are actually
 * re-adjusted with, and because the UF is published every calendar day it
 * answers at day precision. The UF tracks the CPI with the lag it is built
 * with — a day's value follows the CPI of two months earlier — so the
 * variation it reports is inflation as the UF carries it, not the INE's
 * month-on-month figure.
 */
object UfReajuste {

    private val MC = MathContext.DECIMAL64
    private const val DAYS_IN_YEAR = 365.25

    fun convert(amount: BigDecimal, from: DatedValue, to: DatedValue): ReajusteResult {
        require(from.value.signum() > 0 && to.value.signum() > 0) {
            "El valor de la UF debe ser positivo"
        }

        val units = amount.divide(from.value, 4, RoundingMode.HALF_UP)
        val factor = to.value.divide(from.value, MC)
        val days = ChronoUnit.DAYS.between(from.date, to.date)

        val annualised = if (days == 0L) BigDecimal.ZERO else {
            val years = days.toDouble() / DAYS_IN_YEAR
            val a = Math.pow(factor.toDouble(), 1.0 / years) - 1.0
            if (a.isNaN() || a.isInfinite()) BigDecimal.ZERO
            else BigDecimal(a * 100).setScale(2, RoundingMode.HALF_UP)
        }

        return ReajusteResult(
            amount = amount,
            from = from.date,
            to = to.date,
            ufAtFrom = from.value,
            ufAtTo = to.value,
            ufUnits = units,
            // Rounded once, at the end: converting through UF units and then
            // rounding twice would drift on large amounts.
            adjustedAmount = amount.multiply(factor).setScale(0, RoundingMode.HALF_UP),
            factor = factor.setScale(6, RoundingMode.HALF_UP),
            variationPct = (factor - BigDecimal.ONE).multiply(BigDecimal(100))
                .setScale(2, RoundingMode.HALF_UP),
            annualisedPct = annualised,
            days = days,
        )
    }
}
